Prevent incorrect use of `flatMap`

gfontenot opened the pull request:
Swift will happily coerce functions of the type `A -> B` into functions of the
type `A -> B?`. That means that using `flatMap` with a function that doesn't
return an optional will compile. But then at runtime, you will run into weird
behaviour that you don't expect, or worse, Swift will segfault.

In order to prevent these problems from happening in the future, we're going
to throw errors ourselves if `flatMap` is used incorrectly with a message
letting the user know that the type of the function isn't correct. Using a
default (but ultimately unused) implementation of `map` will push people in
the right direction if they go digging.

Credit to @mattrubin for the idea: 
https://github.com/thoughtbot/Argo/issues/34#issuecomment-70166588

gfontenot committed:
Prevent incorrect use of `flatMap`

Swift will happily coerce functions of the type `A -> B` into functions
of the type `A -> B?`. That means that using `flatMap` with a function
that doesn't return an optional will compile. But then at runtime, you
will run into weird behaviour that you don't expect, or worse, Swift
will segfault.

In order to prevent these problems from happening in the future, we're
going to throw errors ourselves if `flatMap` is used incorrectly with a
message letting the user know that the type of the function isn't
correct. Using a default (but ultimately unused) implementation of `map`
will push people in the right direction if they go digging.

Credit to @mattrubin for the idea:
https://github.com/thoughtbot/Argo/issues/34#issuecomment-70166588

tonyd256 commented:
:+1: :+1: :+1: :+1: :fire: :godmode: :tangerine:

lilyball commented:
Swift seems to think it's ok to use a function of type `A -> B` where a function of type `A -> B?` is expected. The only troubles here have to do with bugs in the Swift compiler. Personally, I don't think it makes sense to prohibit this usage because of a bug that will presumably get fixed eventually (and I certainly hope someone's already filed a radar on this).

All that said, using a function `A -> B` here is kind of pointless, since what you really want is just to map it. I would not think it wrong to keep this change but fix the error message to say something like `"function (T -> U) does not return [U], perhaps you meant <^>"`.

gfontenot commented:
@kballard My concern with that is someone misunderstanding that error message and trying to just swap `<^>` for `>>-` and end up with `val <^> func` and an even more opaque error message.

mattrubin commented on Source/Array.swift:
```diff
@@ -40,6 +40,11 @@ public func >>-<T, U>(a: [T], f: T -> [U]) -> [U] {
+@availability(*, unavailable, message="function (T -> U) is not equal to function (T -> [U])")
+public func >>-<T, U>(a: [T], f: T -> U) -> [U] {
+    return a.map(f)
+}
```
Is this necessary for arrays? I don't think Swift will coerce `(T) -> U` to `(T) -> [U]` the way it will with optionals.

gfontenot replied:
No, it doesn't. That would be horrible. But without this, it throws the Optional version of the warning. The Array warning serves to override that one and provide the correct error.

lilyball commented:
@gfontenot How about `"function (T -> U) does not return [U], perhaps you meant f <^> val"`?

gfontenot commented:
Also, if the bug that results in the segfaults was fixed, I'd probably end up removing this protection. I'm less concerned with preventing "incorrect" usage of the operator than I am with protecting users (including myself) from potentially dangerous behavior. If I were looking to force correctness, I'd add a warning about using `apply` with non-optional functions as well.

gfontenot commented:
@kballard Yes, I like that much better. Thanks.

gfontenot committed:
Update error messaging to include note about <^>

We should try to push people towards using `map` here.

credit to @kballard for the wording
